pkg/costattribution/sample_tracker.go

```diff
@@ -0,0 +1,308 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package costattribution
+
+import (
+	"bytes"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+const sep = rune(0x80)
+
+type observation struct {
+	lastUpdate         atomic.Int64
+	receivedSample     atomic.Float64
+	discardedSampleMtx sync.Mutex
+	discardedSample    map[string]*atomic.Float64
+	totalDiscarded     atomic.Float64
+}
+
+type SampleTracker struct {
+	userID                     string
+	labels                     []string
+	maxCardinality             int
+	receivedSamplesAttribution *prometheus.Desc
+	discardedSampleAttribution *prometheus.Desc
+	overflowLabels             []string
+	observed                   map[string]*observation
+	observedMtx                sync.RWMutex
+	overflowSince              atomic.Int64
+	overflowCounter            observation
+	cooldownDuration           time.Duration
+	logger                     log.Logger
+}
+
+func newSampleTracker(userID string, trackedLabels []string, limit int, cooldown time.Duration, logger log.Logger) *SampleTracker {
+	orderedLables := slices.Clone(trackedLabels)
+	slices.Sort(orderedLables)
+
```

Comment:
We can also remove this for same reason.
```suggestion
```

Reply:
it is called from 2 places, I moved the order function to the other caller to avoid double ordering 5c26a0b4dcf3040586b33dafe91fa403c8be0245